imports/lib/schemas/Setting.ts

```diff
@@ -1,92 +1,78 @@
-    value: t.type({ id: t.string }),
-  }),
-  t.type({
-    name: t.literal('discord.bot'),
-    value: t.type({
-      token: t.string,
-    }),
-  }),
-  t.type({
-    name: t.literal('discord.guild'),
-    value: t.type({
-      guild: GuildType,
-    }),
-  }),
-  t.type({
-    name: t.literal('email.branding'),
-    value: t.type({
-      from: t.union([t.string, t.undefined]),
-      enrollAccountMessageSubjectTemplate: t.union([t.string, t.undefined]),
-      enrollAccountMessageTemplate: t.union([t.string, t.undefined]),
-      existingJoinMessageSubjectTemplate: t.union([t.string, t.undefined]),
-      existingJoinMessageTemplate: t.union([t.string, t.undefined]),
-    }),
-  }),
-  t.type({
-    name: t.literal('teamname'),
-    value: t.type({
-      teamName: t.string,
-    }),
-  }),
-  t.type({
-    name: t.literal('google.script'),
-    value: t.type({
-      sharedSecret: t.string,
-      scriptId: t.string,
-      contentHash: t.string,
-      endpointUrl: t.union([t.undefined, t.string]),
-    }),
-  }),
-]);
-export const SettingCodec = t.intersection([
-  BaseCodec,
-  SettingCodecTaggedUnion,
+/* TODO: handle soft deletion */
+
+const SettingDiscriminatedUnion = z.discriminatedUnion('name', [
+  z.object({
+    name: z.literal('gdrive.credential'),
+    value: z.object({
+      refreshToken: z.string(),
+      email: z.string(),
+    }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('gdrive.root'),
+    value: z.object({ id: z.string() }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('gdrive.template.document'),
+    value: z.object({ id: z.string() }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('gdrive.template.spreadsheet'),
+    value: z.object({ id: z.string() }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('discord.bot'),
+    value: z.object({ token: z.string() }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('discord.guild'),
+    value: z.object({
+      guild: z.object({ id: z.string(), name: z.string() }).strict(),
+    }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('email.branding'),
+    value: z.object({
+      from: z.string().optional(),
+      enrollAccountMessageSubjectTemplate: z.string().optional(),
+      enrollAccountMessageTemplate: z.string().optional(),
+      existingJoinMessageSubjectTemplate: z.string().optional(),
+      existingJoinMessageTemplate: z.string().optional(),
+    }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('teamname'),
+    value: z.object({ teamName: z.string() }).strict(),
+  }).strict(),
+  z.object({
+    name: z.literal('google.script'),
+    value: z.object({
+      sharedSecret: z.string(),
+      scriptId: z.string(),
+      contentHash: z.string(),
+      endpointUrl: z.string().optional(),
+    }).strict(),
+  }).strict(),
 ]);
-export type SettingType = t.TypeOf<typeof SettingCodec>;
+const Setting = z.object({
```

Comment:
Ah, I imagine this particular `z.object()` is probably worth extracting as `BaseSchema` and then each of the things that we have that are based on `Base` can just `.and()` in the relevant additional fields.

Reply:
Yeah, I was going to double-back and make something more like `Base`, I just wanted to see if I could get the discriminated union type working. (I am somewhat pleased that it worked with no changes to application code)